optuna/visualization/matplotlib/_optimization_history.py

```diff
@@ -128,11 +130,93 @@ def _get_optimization_history_plot(
         return ax
 
-    ax = _get_optimization_histories(studies, target, target_name, ax)
+    if error_bar:
+        ax = _get_optimization_histories_with_error_bar(studies, target, target_name, ax)
+    else:
+        ax = _get_optimization_histories(studies, target, target_name, ax)
+
     plt.legend(bbox_to_anchor=(1.05, 1.0), loc="upper left")
     return ax
 
 
+def _get_optimization_histories_with_error_bar(
+    studies: List[Study],
+    target: Optional[Callable[[FrozenTrial], float]],
+    target_name: str,
+    ax: "Axes",
+) -> "Axes":
+
+    max_trial_number = np.max(
+        [
+            trial.number
+            for study in studies
+            for trial in study.get_trials(states=(TrialState.COMPLETE,))
+        ]
+    )
+
+    _target: Callable[[FrozenTrial,], float]
+    if target is None:
+
+        def _target(t: FrozenTrial) -> float:
+            return cast(float, t.value)
+
+    else:
+        _target = target
+
+    target_values: List[List[float]] = [[] for _ in range(max_trial_number + 2)]
+    for study in studies:
+        trials = study.get_trials(states=(TrialState.COMPLETE,))
+        for t in trials:
+            target_values[t.number].append(_target(t))
+
+    mean_of_target_values = [np.mean(v) if len(v) > 0 else None for v in target_values]
+    std_of_target_values = [np.std(v) if len(v) > 0 else None for v in target_values]
+    trial_numbers = np.arange(max_trial_number + 2)[[v is not None for v in mean_of_target_values]]
+    means = np.asarray(mean_of_target_values)[trial_numbers]
+    stds = np.asarray(std_of_target_values)[trial_numbers]
+
+    plt.errorbar(
+        x=trial_numbers,
+        y=means,
+        yerr=stds,
+        capsize=5,
+        fmt="o",
+        color="tab:blue",
+    )
+    ax.scatter(trial_numbers, means, color="tab:blue", label=target_name)
+
+
+    if target is None:
+        best_values: List[List[float]] = [[] for _ in range(max_trial_number + 2)]
+        for study in studies:
+            trials = study.get_trials(states=(TrialState.COMPLETE,))
+            if study.direction == StudyDirection.MINIMIZE:
+                best_vs = np.minimum.accumulate([cast(float, t.values) for t in trials])
+            else:
+                best_vs = np.maximum.accumulate([cast(float, t.values) for t in trials])
+
+            for i, t in enumerate(trials):
+                best_values[t.number].append(best_vs[i])
+
+        mean_of_best_values = [np.mean(v) if len(v) > 0 else None for v in best_values]
+        std_of_best_values = [np.std(v) if len(v) > 0 else None for v in best_values]
+        means = np.asarray(mean_of_best_values)[trial_numbers]
+        stds = np.asarray(std_of_best_values)[trial_numbers]
+
+        ax.plot(trial_numbers, means, color="tab:red", label="Best Value")
+        ax.fill_between(
+            x=trial_numbers,
+            y1=np.array(means - stds, float),
+            y2=np.array(means + stds, float),
```

Comment:
I think `means` are `float` values, so we can drop the dtypes.

```suggestion
            y1=np.array(means - stds),
            y2=np.array(means + stds),
```

Reply:
I think it seems to be right. but if case, I get an error like this `
TypeError: ufunc 'isfinite' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''`

Reply:
Thank you for investigating it! Indeed, it comes from filling `None` in `mean_of_target_values = [np.mean(v) if len(v) > 0 else None for v in target_values]`, for example.  So replacing `mean_of_target_values = [np.mean(v) if len(v) > 0 else None for v in target_values]` with  `mean_of_target_values = [np.mean(v) if len(v) > 0 else np.nan for v in target_values]` might be simpler. 

What do you think? cc: @himkt